// react-film/lib/patchStyleOptions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = normalizeStyleOptions;
var DOT_BOX_SIZE = 20;
var DOT_SIZE = 6;
var FLIPPER_BOX_WIDTH = 60;
var FLIPPER_SIZE = 40;
var SCROLL_BAR_HEIGHT = 8;
var SCROLL_BAR_MARGIN = 4;

function normalizeStyleOptions() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      autoCenter = _ref.autoCenter,
      autoHide = _ref.autoHide,
      autoHideFlipperOnEdge = _ref.autoHideFlipperOnEdge,
      cursor = _ref.cursor,
      dir = _ref.dir,
      dotBoxSize = _ref.dotBoxSize,
      dotSize = _ref.dotSize,
      flipperBlurFocusOnClick = _ref.flipperBlurFocusOnClick,
      flipperBoxWidth = _ref.flipperBoxWidth,
      flipperSize = _ref.flipperSize,
      leftFlipperText = _ref.leftFlipperText,
      rightFlipperText = _ref.rightFlipperText,
      scrollBarHeight = _ref.scrollBarHeight,
      scrollBarMargin = _ref.scrollBarMargin,
      showDots = _ref.showDots,
      showFlipper = _ref.showFlipper,
      showScrollBar = _ref.showScrollBar;

  var rtl = dir === 'rtl';
  return {
    autoCenter: autoCenter !== false,
    autoHide: autoHide !== false,
    autoHideFlipperOnEdge: autoHideFlipperOnEdge !== false,
    cursor: typeof cursor === 'undefined' ? 'pointer' : cursor,
    dotBoxSize: typeof dotBoxSize === 'number' ? dotBoxSize : DOT_BOX_SIZE,
    dotSize: typeof dotSize === 'number' ? dotSize : DOT_SIZE,
    flipperBlurFocusOnClick: flipperBlurFocusOnClick === true,
    flipperBoxWidth: typeof flipperBoxWidth === 'number' ? flipperBoxWidth : FLIPPER_BOX_WIDTH,
    flipperSize: typeof flipperSize === 'number' ? flipperSize : FLIPPER_SIZE,
    leftFlipperText: leftFlipperText || (rtl ? '>' : '<'),
    rightFlipperText: rightFlipperText || (rtl ? '<' : '>'),
    scrollBarHeight: typeof scrollBarHeight === 'number' ? scrollBarHeight : SCROLL_BAR_HEIGHT,
    scrollBarMargin: typeof scrollBarMargin === 'number' ? scrollBarMargin : SCROLL_BAR_MARGIN,
    showDots: showDots !== false,
    showFlipper: showFlipper !== false,
    showScrollBar: showScrollBar !== false
  };
}
