// react-film/lib/ReactFilm.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _BasicFilm = _interopRequireDefault(require("./BasicFilm"));

var _Composer = _interopRequireDefault(require("./Composer"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var ReactFilm = function ReactFilm(_ref) {
  var autoCenter = _ref.autoCenter,
      autoHide = _ref.autoHide,
      autoHideFlipperOnEdge = _ref.autoHideFlipperOnEdge,
      children = _ref.children,
      className = _ref.className,
      dir = _ref.dir,
      flipperBlurFocusOnClick = _ref.flipperBlurFocusOnClick,
      height = _ref.height,
      leftFlipperText = _ref.leftFlipperText,
      nonce = _ref.nonce,
      rightFlipperText = _ref.rightFlipperText,
      showDots = _ref.showDots,
      showFlipper = _ref.showFlipper,
      showScrollBar = _ref.showScrollBar,
      styleSet = _ref.styleSet;
  var styleOptions = (0, _react.useMemo)(function () {
    return {
      autoCenter: autoCenter,
      autoHide: autoHide,
      autoHideFlipperOnEdge: autoHideFlipperOnEdge,
      dir: dir,
      leftFlipperText: leftFlipperText,
      flipperBlurFocusOnClick: flipperBlurFocusOnClick,
      rightFlipperText: rightFlipperText,
      showDots: showDots,
      showFlipper: showFlipper,
      showScrollBar: showScrollBar
    };
  }, [autoCenter, autoHide, autoHideFlipperOnEdge, dir, leftFlipperText, flipperBlurFocusOnClick, rightFlipperText, showDots, showFlipper, showScrollBar]);
  return /*#__PURE__*/_react.default.createElement(_Composer.default, {
    dir: dir,
    height: height,
    nonce: nonce,
    numItems: children ? _react.Children.count(children) : 0,
    styleOptions: styleOptions,
    styleSet: styleSet
  }, /*#__PURE__*/_react.default.createElement(_BasicFilm.default, {
    className: className
  }, children));
};

ReactFilm.defaultProps = _objectSpread(_objectSpread({}, _BasicFilm.default.defaultProps), {}, {
  autoCenter: undefined,
  autoHide: undefined,
  autoHideFlipperOnEdge: undefined,
  children: undefined,
  className: undefined,
  dir: undefined,
  flipperBlurFocusOnClick: undefined,
  leftFlipperText: undefined,
  nonce: undefined,
  rightFlipperText: undefined,
  showDots: undefined,
  showFlipper: undefined,
  showScrollBar: undefined,
  styleSet: undefined
});
ReactFilm.propTypes = _objectSpread(_objectSpread({}, _BasicFilm.default.propTypes), {}, {
  autoCenter: _propTypes.default.bool,
  autoHide: _propTypes.default.bool,
  autoHideFlipperOnEdge: _propTypes.default.bool,
  children: _propTypes.default.any,
  className: _propTypes.default.string,
  dir: _propTypes.default.oneOf(['auto', 'ltr', 'rtl']),
  flipperBlurFocusOnClick: _propTypes.default.bool,
  height: _propTypes.default.number,
  leftFlipperText: _propTypes.default.string,
  nonce: _propTypes.default.string,
  rightFlipperText: _propTypes.default.string,
  showDots: _propTypes.default.bool,
  showFlipper: _propTypes.default.bool,
  showScrollBar: _propTypes.default.bool,
  styleSet: _propTypes.default.any
});
var _default = ReactFilm;
exports.default = _default;
