// react-film/lib/ScrollBar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireDefault(require("react"));

var browser = _interopRequireWildcard(require("./browser"));

var _useDir3 = _interopRequireDefault(require("./hooks/useDir"));

var _useScrollBarPercentage = _interopRequireDefault(require("./hooks/useScrollBarPercentage"));

var _useScrollBarWidth3 = _interopRequireDefault(require("./hooks/useScrollBarWidth"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var ScrollBar = function ScrollBar() {
  var _useDir = (0, _useDir3.default)(),
      _useDir2 = (0, _slicedToArray2.default)(_useDir, 1),
      dir = _useDir2[0];

  var _useScrollBarPercenta = (0, _useScrollBarPercentage.default)(),
      _useScrollBarPercenta2 = (0, _slicedToArray2.default)(_useScrollBarPercenta, 1),
      scrollBarPercentage = _useScrollBarPercenta2[0];

  var _useScrollBarWidth = (0, _useScrollBarWidth3.default)(),
      _useScrollBarWidth2 = (0, _slicedToArray2.default)(_useScrollBarWidth, 1),
      scrollBarWidth = _useScrollBarWidth2[0];

  return /*#__PURE__*/_react.default.createElement("div", {
    className: "react-film__scroll-bar react-film__main__slider react-film__main__slider--bottom"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "react-film__scroll-bar__handle",
    style: _objectSpread(_objectSpread({}, dir === 'rtl' ? browser.firefox || browser.safari ? {
      marginRight: "".concat((parseFloat(scrollBarWidth) / 100 - 1) * parseFloat(scrollBarPercentage), "%")
    } : browser.edgeUWP || browser.internetExplorer ? {
      marginRight: "".concat((1 - parseFloat(scrollBarWidth) / 100) * parseFloat(scrollBarPercentage), "%")
    } : {
      marginRight: "".concat((1 - parseFloat(scrollBarWidth) / 100) * (100 - parseFloat(scrollBarPercentage)), "%")
    } : {
      marginLeft: "".concat((1 - parseFloat(scrollBarWidth) / 100) * parseFloat(scrollBarPercentage), "%")
    }), {}, {
      width: scrollBarWidth
    })
  }));
};

var _default = ScrollBar;
exports.default = _default;
